SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/SpatialPingComponent.h"
+
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "TimerManager.h"
+#include "Engine/World.h"
+
+USpatialPingComponent::USpatialPingComponent(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+
+	bReplicates = true;
+
+	bStartWithPingEnabled = true;
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(USpatialPingComponent, ReplicatedPingID);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	// Attempt to cast the owner of this component to the PlayerController class.
+	// Component will do nothing if cast fails.
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (bStartWithPingEnabled)
+	{
+		SetPingEnabled(true);
+	}
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	Super::EndPlay(EndPlayReason);
+	// Clear up the timer before the component is disabled or destroyed.
+	SetPingEnabled(false);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	if (OwningController != nullptr)
+	{
+		// Only execute on owning local client.
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled)
+			{
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled)
+			{
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RoundTripPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		LastSentPingID = 0;
+		// (Re)initialize TArrays
+		SentPingIDs.Init(0, 0);
```

Comment:
```suggestion
		SentPingIDs.Empty();
```

Reply:
These arrays have been replaced so no longer need to worry about emptying them.